File: src/game/landmark-registry.ts
// === Landmark Registry — universal debug listing of macroworld points ===
//
// Data-driven registry of landmark sources. Anything in the world that
// has a (x, y) location and can be enumerated from GameState plugs in
// here with one entry. The DebugOverlay reads this registry to render
// a unified, expandable landmark list.
//
// Adding a new landmark kind = one `registerLandmarkSource()` call.
// No DebugOverlay changes required.

import type {GameState} from './state';

export type LandmarkEntry = {
	/** Unique within its source. */
	id: string;
	name: string;
	x: number;
	y: number;
	/** Optional secondary line (e.g. status, faction). */
	detail?: string;
};

export type LandmarkSource = {
	/** Display label for the kind ('City', 'Village', ...). */
	kind: string;
	/** Tailwind text color class for the entries of this source. */
	color: string;
	/** Pull all current landmarks of this kind from game state. */
	collect(state: GameState): LandmarkEntry[];
};

export const LANDMARK_SOURCES: LandmarkSource[] = [];

export function registerLandmarkSource(source: LandmarkSource): void {
	LANDMARK_SOURCES.push(source);
}

// ── Built-in sources ──────────────────────────────────────────────

registerLandmarkSource({
	kind: 'City',
	color: 'text-amber-300',
	collect: state => state.settlements.map(s => ({
		id: `city_${s.id}`,
		name: s.name,
		x: s.x,
		y: s.y,
		detail: `pop ${s.population} · ${s.mood}`,
	})),
});

registerLandmarkSource({
	kind: 'Village',
	color: 'text-lime-300',
	collect: state => state.villages.map(v => ({
		id: `village_${v.id}`,
		name: v.name,
		x: v.x,
		y: v.y,
		detail: `pop ${v.population}`,
	})),
});

registerLandmarkSource({
	kind: 'Spire',
	color: 'text-purple-300',
	collect: state => state.spires.map(s => ({
		id: `spire_${s.id}`,
		name: `Spire of ${s.spellId}`,
		x: s.x,
		y: s.y,
		detail: s.depleted ? 'depleted' : 'active',
	})),
});

registerLandmarkSource({
	kind: 'Marker',
	color: 'text-cyan-300',
	collect: state => state.markers.map(m => ({
		id: `marker_${m.id}`,
		name: m.label ?? m.id,
		x: m.x,
		y: m.y,
		detail: m.style,
	})),
});

File: timaert_c/src/macro/landmark_registry.cpp
#include "macro/landmark_registry.h"
#include "macro/state.h"
#include "macro/markers.h"
#include <cstdio>

namespace sm {

namespace {

// Match TS Tailwind class colours from landmark-registry.ts as ARGB.
// text-amber-300, text-lime-300, text-purple-300, text-cyan-300.
constexpr std::uint32_t kColorCity    = 0xFFFCD34Du; // amber-300
constexpr std::uint32_t kColorVillage = 0xFFBEF264u; // lime-300
constexpr std::uint32_t kColorSpire   = 0xFFD8B4FEu; // purple-300
constexpr std::uint32_t kColorMarker  = 0xFF67E8F9u; // cyan-300

const char* mood_label(SettlementMood m) {
    switch (m) {
        case SettlementMood::Prosperous: return "prosperous";
        case SettlementMood::Stable:     return "stable";
        case SettlementMood::Tense:      return "tense";
        case SettlementMood::Unrest:     return "unrest";
        case SettlementMood::Revolt:     return "revolt";
    }
    return "?";
}

const char* marker_kind(MarkerStyle s) {
    switch (s) {
        case MarkerStyle::Quest:    return "quest";
        case MarkerStyle::POI:      return "poi";
        case MarkerStyle::Danger:   return "danger";
        case MarkerStyle::Waypoint: return "waypoint";
    }
    return "?";
}

} // namespace

std::vector<LandmarkEntry> collect_landmarks(const GameState& gs) {
    std::vector<LandmarkEntry> out;
    out.reserve(gs.settlements.size() + gs.villages.size()
              + gs.spires.size() + gs.markers.size());

    char buf[64];

    for (const auto& s : gs.settlements) {
        std::snprintf(buf, sizeof(buf), "pop %d \xC2\xB7 %s",
                      s.population, mood_label(s.mood));
        out.push_back({"City",
                       "city_" + std::to_string(s.id),
                       s.name, buf, s.x, s.y, kColorCity});
    }
    for (const auto& v : gs.villages) {
        std::snprintf(buf, sizeof(buf), "pop %d", v.population);
        out.push_back({"Village",
                       "village_" + std::to_string(v.id),
                       v.name, buf, v.x, v.y, kColorVillage});
    }
    for (const auto& sp : gs.spires) {
        std::snprintf(buf, sizeof(buf), "spell 0x%08X",
                      static_cast<unsigned>(sp.spellId));
        out.push_back({"Spire",
                       "spire_" + std::to_string(sp.id),
                       sp.depleted ? "Depleted Spire" : "Spire",
                       sp.depleted ? "depleted" : "active",
                       sp.x, sp.y, kColorSpire});
        (void)buf;
    }
    for (const auto& m : gs.markers) {
        out.push_back({"Marker",
                       "marker_" + m.id,
                       m.label.empty() ? m.id : m.label,
                       marker_kind(m.style),
                       int(m.x), int(m.y), kColorMarker});
    }

    return out;
}

} // namespace sm
